[Clang][CodeGen] Remove explicit insertion of AllocToken pass

melver opened the pull request:
Remove explicit insertion of the AllocTokenPass, which is now handled by
the PassBuilder. Emit AllocToken configuration as LLVM module flags to
persist into the backend.

Specifically, this also means it will now be handled by LTO backend
phases; this avoids interference with other optimizations (e.g. PGHO)
and enable late heap-allocation optimizations with LTO enabled.

---

This change is part of the following series:
1. https://github.com/llvm/llvm-project/pull/169242
2. https://github.com/llvm/llvm-project/pull/169243
3. https://github.com/llvm/llvm-project/pull/169358
4. https://github.com/llvm/llvm-project/pull/169360

melver committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.8-beta.1

melver committed:
fix formatting

Created using spr 1.3.8-beta.1

teresajohnson reviewed (commented):
Rather than having different ways to inject this pass for the different LTO backend pipelines, and for distributed vs in process ThinLTO, and also rather than passing internal options to the linker driver which we're trying to get away from, can the pass be added unconditionally from the PassBuilder::build*LTODefaultPipeline in the correct location and guarded by something in the IR? In fact, looking at the AllocToken pass, it already appears to be a noop without the necessary attributes and intrinsics. Can the mode also be passed down in the IR via an attribute or at least a module flag to simplify the enablement?

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rework

Created using spr 1.3.8-beta.1

melver commented:
> Rather than having different ways to inject this pass for the different LTO backend pipelines, and for distributed vs in process ThinLTO, and also rather than passing internal options to the linker driver which we're trying to get away from, can the pass be added unconditionally from the PassBuilder::build*LTODefaultPipeline in the correct location and guarded by something in the IR?

I reworked it and it's added in PassBuilder now.

> In fact, looking at the AllocToken pass, it already appears to be a noop without the necessary attributes and intrinsics. Can the mode also be passed down in the IR via an attribute or at least a module flag to simplify the enablement?

Reworked to use module flags.

PTAL. Thanks!

melver committed:
fix test

Created using spr 1.3.8-beta.1

melver committed:
fix test

Created using spr 1.3.8-beta.1

teresajohnson reviewed (approved):
lgtm

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
remove addition of pass by clang completely

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.8-beta.1

[skip ci]

melver committed:
rebase

Created using spr 1.3.8-beta.1

melver committed:
rebase

Created using spr 1.3.8-beta.1

dyung commented:
Hi @melver, one of the tests you added `clang/test/CodeGen/distributed-thin-lto/memprof-pgho.cpp` seems to be failing on a bot, can you take a look? (You might not have been notified since the bot was already failing with another test failure which has since been fixed)

https://lab.llvm.org/buildbot/#/builders/190/builds/31993

melver commented:
> clang/test/CodeGen/distributed-thin-lto/memprof-pgho.cpp

Sorry about that - it looks like this needs an x86 build. This should fix it: https://github.com/llvm/llvm-project/pull/170321

This is similar to the other tests in `distributed-thin-lto` which have the REQUIRES line.
